src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -366,60 +379,37 @@ result<json::Document> parse_json(std::string_view v) {
-        schema_stream, metaschema_doc};
-
-    // then parse schema to json
-    auto schema_json = json::Document{};
-    schema_json.Populate(validating_reader);
-
-    if (auto parse_res = validating_reader.GetParseResult();
-        parse_res.IsError()) {
-        // schema_json is either not a json document
-        // or it's not a valid json according to metaschema
-
-        // Check the validation result
-        if (!validating_reader.IsValid()) {
-            // not a valid schema draft4 according to metaschema. retrieve some
-            // info and return error
-            auto error_loc_metaschema = json::StringBuffer{};
-            auto error_loc_schema = json::StringBuffer{};
-            validating_reader.GetInvalidSchemaPointer().StringifyUriFragment(
-              error_loc_metaschema);
-            validating_reader.GetInvalidDocumentPointer().StringifyUriFragment(
-              error_loc_schema);
-            auto invalid_keyword = validating_reader.GetInvalidSchemaKeyword();
-
-            return error_info{
-              error_code::schema_invalid,
-              fmt::format(
-                "Invalid json schema: '{}', invalid metaschema: '{}', invalid "
-                "keyword: '{}'",
-                std::string_view{
-                  error_loc_schema.GetString(), error_loc_schema.GetLength()},
-                std::string_view{
-                  error_loc_metaschema.GetString(),
-                  error_loc_metaschema.GetLength()},
-                invalid_keyword)};
-        } else {
-            // not a valid json document, return error
-            return error_info{
-              error_code::schema_invalid,
-              fmt::format(
-                "Malformed json schema: {} at offset {}",
-                rapidjson::GetParseError_En(parse_res.Code()),
-                parse_res.Offset())};
-        }
+    auto validator = json::SchemaValidator{metaschema_doc};
+
+    if (!schema.Accept(validator)) {
+        // not a valid schema draft4 according to metaschema. retrieve some
+        // info and return error
+
+        auto error_loc_metaschema = json::StringBuffer{};
+        auto error_loc_schema = json::StringBuffer{};
+        validator.GetInvalidSchemaPointer().StringifyUriFragment(
+          error_loc_metaschema);
+        validator.GetInvalidDocumentPointer().StringifyUriFragment(
+          error_loc_schema);
+        auto invalid_keyword = validator.GetInvalidSchemaKeyword();
+
+        return error_info{
+          error_code::schema_invalid,
+          fmt::format(
+            "Invalid json schema: '{}', invalid metaschema: '{}', invalid "
+            "keyword: '{}'",
+            std::string_view{
+              error_loc_schema.GetString(), error_loc_schema.GetLength()},
+            std::string_view{
+              error_loc_metaschema.GetString(),
+              error_loc_metaschema.GetLength()},
+            invalid_keyword)};
```

Comment:
This seems mostly a duplicate of [above](https://github.com/redpanda-data/redpanda/pull/20817/commits/879718043ba82cccffeaf5e9186ee8a23d90db7b#diff-a401c2c6d951b2e78b32cb570bba6b9d26282815e93186591a1b82b0db33efdbR306-R324)

Reply:
intermediate commit to try to get a cleaner history, the final state does not have this duplication